server/sonar-server/src/main/java/org/sonar/server/component/index/ComponentIndexer.java

```diff
@@ -65,13 +68,24 @@ public void index() {
     doIndexByProjectUuid(null);
   }
 
-  /**
-   * Update the index for one specific project. The current data from the database is used.
-   */
-  public void indexByProjectUuid(String projectUuid) {
-    requireNonNull(projectUuid);
-    deleteComponentsByProjectUuid(projectUuid);
-    doIndexByProjectUuid(projectUuid);
+  @Override
+  public void indexProject(String projectUuid, Cause cause) {
+    switch (cause) {
+      case PROJECT_CREATION:
+      case PROJECT_KEY_UPDATE:
+      case NEW_ANALYSIS:
+        deleteProject(projectUuid);
+        doIndexByProjectUuid(projectUuid);
+        break;
+      default:
+        // defensive case
+        throw new IllegalStateException("Unsupported cause: " + cause);
+    }
```

Comment:
no because:
* cause is not specific to an index. Only indexers know if re-indexing is needed.
* the default switch is a safeguard. It ensures that all types of cause are handled by indexers. Adding a new Cause value will have to be handled everywhere.